oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/JdbcOAuth2AuthorizedClientService.java

```diff
@@ -0,0 +1,516 @@
+			OAuth2AuthorizedClient authorizedClient = authorizedClientHolder.getAuthorizedClient();
+			Authentication principal = authorizedClientHolder.getPrincipal();
+			ClientRegistration clientRegistration = authorizedClient.getClientRegistration();
+			OAuth2AccessToken accessToken = authorizedClient.getAccessToken();
+			OAuth2RefreshToken refreshToken = authorizedClient.getRefreshToken();
+
+			OAuth2AuthorizedClientData authorizedClientData = new OAuth2AuthorizedClientData();
+			authorizedClientData.setClientRegistrationId(clientRegistration.getRegistrationId());
+			authorizedClientData.setPrincipalName(principal.getName());
+			authorizedClientData.setAccessTokenType(accessToken.getTokenType().getValue());
+			authorizedClientData.setAccessTokenValue(accessToken.getTokenValue());
+			authorizedClientData.setAccessTokenIssuedAt(Timestamp.from(accessToken.getIssuedAt()));
+			authorizedClientData.setAccessTokenExpiresAt(Timestamp.from(accessToken.getExpiresAt()));
+			if (!CollectionUtils.isEmpty(accessToken.getScopes())) {
+				authorizedClientData.setAccessTokenScopes(StringUtils.collectionToDelimitedString(accessToken.getScopes(), ","));
+			}
+			if (refreshToken != null) {
+				authorizedClientData.setRefreshTokenValue(refreshToken.getTokenValue());
+				if (refreshToken.getIssuedAt() != null) {
+					authorizedClientData.setRefreshTokenIssuedAt(Timestamp.from(refreshToken.getIssuedAt()));
+				}
+			}
+			return authorizedClientData;
+		}
+	}
+
+	/**
+	 * A holder for an {@link OAuth2AuthorizedClient} and End-User {@link Authentication} (Resource Owner).
+	 */
+	public static final class OAuth2AuthorizedClientHolder {
+		private final OAuth2AuthorizedClient authorizedClient;
+		private final Authentication principal;
+
+		/**
+		 * Constructs an {@code OAuth2AuthorizedClientHolder} using the provided parameters.
+		 *
+		 * @param authorizedClient the authorized client
+		 * @param principal the End-User {@link Authentication} (Resource Owner)
+		 */
+		public OAuth2AuthorizedClientHolder(OAuth2AuthorizedClient authorizedClient, Authentication principal) {
+			Assert.notNull(authorizedClient, "authorizedClient cannot be null");
+			Assert.notNull(principal, "principal cannot be null");
+			this.authorizedClient = authorizedClient;
+			this.principal = principal;
+		}
+
+		/**
+		 * Returns the {@link OAuth2AuthorizedClient}.
+		 *
+		 * @return the {@link OAuth2AuthorizedClient}
+		 */
+		public OAuth2AuthorizedClient getAuthorizedClient() {
+			return this.authorizedClient;
+		}
+
+		/**
+		 * Returns the End-User {@link Authentication} (Resource Owner).
+		 *
+		 * @return the End-User {@link Authentication} (Resource Owner)
+		 */
+		public Authentication getPrincipal() {
+			return this.principal;
+		}
+	}
+
+	/**
+	 * The data (entity) representation of an {@link OAuth2AuthorizedClient}.
+	 */
+	public static class OAuth2AuthorizedClientData {
```

Comment:
What is the need for this type vs allowing the user to inject a `RowMapper` that produces an `OAuth2AuthorizedClient`?

Reply:
@rwinch The `OAuth2AuthorizedClientData` is an entity representation that IMO is simpler to deal with compared to returning an `OAuth2AuthorizedClient`. The logic in `OAuth2AuthorizedClientDataConverter` is a bit involved and would require a bit too much work for the user to implement. Instead they can simply reuse `OAuth2AuthorizedClientDataConverter` and implement a `RowMapper<OAuth2AuthorizedClientData>` that performs some data translation/conversion (if necessary), eg. decoding/decrypting the access/refresh token.

Reply:
Take a look at the test `JdbcOAuth2AuthorizedClientServiceTests.tableDefinitionWhenCustomThenAbleToOverride`. It demonstrates what the user would need to implement if they preferred to change the table definition from the standard definition it expects. The `RowMapper<OAuth2AuthorizedClientData>` needs to be implemented but both `Converter`'s are reused.

Reply:
I don't like having this additional type. If `OAuth2AuthorizedClient` is difficult to use, we should just make that easier vs creating another type.

Reply:
I removed `OAuth2AuthorizedClientData`